# Data Cleaning for Segmenting credit card customers using K-Means in Python - Guided Project

Doing a guided project where I do not have access to the clean data set used for the project. However, I was fortune to find a very similar data on Kaggle "https://www.kaggle.com/datasets/sakshigoyal7/credit-card-customers?resource=download" which was downloaded on the 26/05/2025. So, didn't a side project to clean the data to come close to the data set used for the guided projected.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
BankChurners_df = pd.read_csv('BankChurners.csv')
BankChurners_df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

In [3]:
BankChurners_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  object 
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                           

In [4]:
credit_card_df =  BankChurners_df[['CLIENTNUM', 'Customer_Age',	'Gender', 'Dependent_count', 'Education_Level',	
                                   'Marital_Status', 'Income_Category', 'Months_on_book','Total_Relationship_Count',
                                   'Months_Inactive_12_mon', 'Credit_Limit', 'Total_Trans_Amt',	'Total_Trans_Ct', 'Avg_Utilization_Ratio'
                                   
]]

In [5]:
credit_card_df.head()

CLIENTNUM  Customer_Age Gender  Dependent_count Education_Level  \
0  768805383            45      M                3     High School   
1  818770008            49      F                5        Graduate   
2  713982108            51      M                3        Graduate   
3  769911858            40      F                4     High School   
4  709106358            40      M                3      Uneducated   

  Marital_Status Income_Category  Months_on_book  Total_Relationship_Count  \
0        Married     $60K - $80K              39                         5   
1         Single  Less than $40K              44                         6   
2        Married    $80K - $120K              36                         4   
3        Unknown  Less than $40K              34                         3   
4        Married     $60K - $80K              21                         5   

   Months_Inactive_12_mon  Credit_Limit  Total_Trans_Amt  Total_Trans_Ct  \
0                       1       12691.0             1144              42   
1                       1        8256.0             1291              33   
2                       1        3418.0             1887              20   
3                       4        3313.0             1171              20   
4                       1        4716.0              816              28   

   Avg_Utilization_Ratio  
0                  0.061  
1                  0.105  
2                  0.000  
3                  0.760  
4                  0.000

In [6]:
credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           10127 non-null  object 
 5   Marital_Status            10127 non-null  object 
 6   Income_Category           10127 non-null  object 
 7   Months_on_book            10127 non-null  int64  
 8   Total_Relationship_Count  10127 non-null  int64  
 9   Months_Inactive_12_mon    10127 non-null  int64  
 10  Credit_Limit              10127 non-null  float64
 11  Total_Trans_Amt           10127 non-null  int64  
 12  Total_Trans_Ct            10127 non-null  int64  
 13  Avg_Utilization_Ratio     10127 non-null  float64
dtypes: flo

Based on the above, there are non missing data. However from experience object data types sometimes carries unwanted value

In [7]:
# Breakdown the Gender to see the values it holds
credit_card_df['Gender'].value_counts()

Gender
F    5358
M    4769
Name: count, dtype: int64

In [8]:
# Breakdown the Education level to see the values it holds
credit_card_df['Education_Level'].value_counts()

Education_Level
Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: count, dtype: int64

Comparing my data set to the one used in the project, there are no unknown values for Education Level.
<br> The different between my data and data quest data
| Education Level | My Data | Data quest Data | Different |
|-----------------|---------|-----------------|-----------|
| Graduate        | 3128    | 3685            | - 557     |
| High School     | 2013    | 2351            | - 338     |
| Uneducated      | 1487    | 1755            | - 268     |
| College         | 1013    | 1192            | - 179     |
| Post Graduate   | 516     | 616             | - 100     |
| Doctorate       | 451     | 528             | - 77      |

<br> The sum  of the all the different makes the total of my unknown 1519

In [9]:
# Breakdown the Marital Status to see the values it holds
credit_card_df['Marital_Status'].value_counts()

Marital_Status
Married     4687
Single      3943
Unknown      749
Divorced     748
Name: count, dtype: int64

In [10]:
# Breakdown the Income_Category to see the values it holds
credit_card_df['Income_Category'].value_counts()

Income_Category
Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: count, dtype: int64

Dataquest data doesn't have Income category, it however have estimated income, without any unknowns.

## Data Cleaning process

### Cleaning the Education Level

In [11]:
# Analyze records where Education_Level is known,  
# grouping by Income_Category and Gender to examine distribution.  
# Count occurrences, reset index for clarity, and sort by Income_Category.  

(credit_card_df[credit_card_df['Education_Level']!='Unknown'][['Income_Category', 'Gender']]
 .value_counts().reset_index().sort_values(by=['Income_Category']))

Income_Category Gender  count
6         $120K +      M    607
4     $40K - $60K      F    859
5     $40K - $60K      M    662
2     $60K - $80K      M   1192
1    $80K - $120K      M   1316
0  Less than $40K      F   2767
7  Less than $40K      M    238
3         Unknown      F    920
8         Unknown      M     47

In [12]:
# Analyze records where Education_Level is 'Unknown',  
# grouping by Income_Category and Gender to understand distribution.  
# Count occurrences, reset index for readability, and sort by Income_Category.  

(credit_card_df[credit_card_df['Education_Level']=='Unknown'][['Income_Category', 'Gender']]
 .value_counts().reset_index().sort_values(by=['Income_Category']))

Income_Category Gender  count
5         $120K +      M    120
3     $40K - $60K      F    155
6     $40K - $60K      M    114
2     $60K - $80K      M    210
1    $80K - $120K      M    219
0  Less than $40K      F    517
7  Less than $40K      M     39
4         Unknown      F    140
8         Unknown      M      5

After observing the data and the two table above, it was concluded that the data features that might provide in-depth information on a customer level of education income category and gender. 

In [13]:
# Estimating education level distribution for unknown cases based on Income Category and Gender  

# Copy the original dataframe to preserve raw data  
credit_clean = credit_card_df.copy()

# Define income categories and gender groups for iteration  
Income_Cat = ['$120K +', '$80K - $120K', '$60K - $80K', '$40K - $60K', 'Less than $40K', 'Unknown'] 
Genders = ['M', 'F'] 

# List to store processed customer data  
all_customer_data = []

# Iterate over each gender 
for Gender in Genders: # Outer loop for Gender
    for Income in Income_Cat: # Inner loop for Income Category
        # Filter for customers with known education levels within the current gender and income category
        filtered_data = credit_clean[
            (credit_clean['Income_Category'] == Income) &
            (credit_clean['Education_Level'] != 'Unknown') &
            (credit_clean['Gender'] == Gender) # Use the Gender from the outer loop
        ]

        # Count total customers in the filtered subset 
        total_customer = filtered_data.shape[0]

        # Proceed only if there are customers in the filtered subset
        if total_customer > 0: 
            # Count occurrences of each education level
            
            data = filtered_data['Education_Level'].value_counts().reset_index()
            data.columns = ['Education_Level', 'count']

            # Assign corresponding income category and gender
            data['Income_Category'] = Income
            data['Gender'] = Gender 

            # Compute client ratio based on total customers in the category
            data['client_ratio'] = data['count'] / total_customer
            
            # Store processed data 
            all_customer_data.append(data)
        else:
            # Inform user when no matching customers exist
            print(f"No {Gender} customer with known education level found for Income Category: {Income}")

# Combine all collected data into a single dataframe 
customer = pd.concat(all_customer_data, ignore_index=True)

# Display the final dataset 
customer

No F customer with known education level found for Income Category: $120K +
No F customer with known education level found for Income Category: $80K - $120K
No F customer with known education level found for Income Category: $60K - $80K


Education_Level  count Income_Category Gender  client_ratio
0         Graduate    204         $120K +      M      0.336079
1      High School    147         $120K +      M      0.242175
2       Uneducated    119         $120K +      M      0.196046
3          College     70         $120K +      M      0.115321
4        Doctorate     37         $120K +      M      0.060956
5    Post-Graduate     30         $120K +      M      0.049423
6         Graduate    478    $80K - $120K      M      0.363222
7      High School    308    $80K - $120K      M      0.234043
8       Uneducated    217    $80K - $120K      M      0.164894
9          College    175    $80K - $120K      M      0.132979
10   Post-Graduate     81    $80K - $120K      M      0.061550
11       Doctorate     57    $80K - $120K      M      0.043313
12        Graduate    422     $60K - $80K      M      0.354027
13     High School    307     $60K - $80K      M      0.257550
14      Uneducated    195     $60K - $80K      M      0.163591
15         College    132     $60K - $80K      M      0.110738
16   Post-Graduate     77     $60K - $80K      M      0.064597
17       Doctorate     59     $60K - $80K      M      0.049497
18        Graduate    234     $40K - $60K      M      0.353474
19     High School    160     $40K - $60K      M      0.241692
20      Uneducated    116     $40K - $60K      M      0.175227
21         College     75     $40K - $60K      M      0.113293
22   Post-Graduate     50     $40K - $60K      M      0.075529
23       Doctorate     27     $40K - $60K      M      0.040785
24        Graduate    100  Less than $40K      M      0.420168
25     High School     54  Less than $40K      M      0.226891
26      Uneducated     37  Less than $40K      M      0.155462
27         College     26  Less than $40K      M      0.109244
28       Doctorate     11  Less than $40K      M      0.046218
29   Post-Graduate     10  Less than $40K      M      0.042017
30        Graduate     20         Unknown      M      0.425532
31     High School      9         Unknown      M      0.191489
32      Uneducated      7         Unknown      M      0.148936
33   Post-Graduate      5         Unknown      M      0.106383
34         College      3         Unknown      M      0.063830
35       Doctorate      3         Unknown      M      0.063830
36        Graduate    319     $40K - $60K      F      0.371362
37     High School    195     $40K - $60K      F      0.227008
38      Uneducated    133     $40K - $60K      F      0.154831
39         College    108     $40K - $60K      F      0.125728
40   Post-Graduate     61     $40K - $60K      F      0.071013
41       Doctorate     43     $40K - $60K      F      0.050058
42        Graduate   1039  Less than $40K      F      0.375497
43     High School    617  Less than $40K      F      0.222985
44      Uneducated    485  Less than $40K      F      0.175280
45         College    319  Less than $40K      F      0.115287
46   Post-Graduate    160  Less than $40K      F      0.057824
47       Doctorate    147  Less than $40K      F      0.053126
48        Graduate    312         Unknown      F      0.339130
49     High School    216         Unknown      F      0.234783
50      Uneducated    178         Unknown      F      0.193478
51         College    105         Unknown      F      0.114130
52       Doctorate     67         Unknown      F      0.072826
53   Post-Graduate     42         Unknown      F      0.045652

In [14]:
# Convert 'Unknown' values to NaN to facilitate missing value handling and iteration
credit_clean['Education_Level'].replace('Unknown', np.NaN, inplace = True)
credit_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           8608 non-null   object 
 5   Marital_Status            10127 non-null  object 
 6   Income_Category           10127 non-null  object 
 7   Months_on_book            10127 non-null  int64  
 8   Total_Relationship_Count  10127 non-null  int64  
 9   Months_Inactive_12_mon    10127 non-null  int64  
 10  Credit_Limit              10127 non-null  float64
 11  Total_Trans_Amt           10127 non-null  int64  
 12  Total_Trans_Ct            10127 non-null  int64  
 13  Avg_Utilization_Ratio     10127 non-null  float64
dtypes: flo

In [15]:
# Imputing missing Education_Level values using probabilities derived from known data  

# Define income categories and gender groups for iteration
Income_Cat = ['$120K +', '$80K - $120K', '$60K - $80K', '$40K - $60K', 'Less than $40K', 'Unknown']
Genders = ['M', 'F']

# Iterate through gender and income category combinations 
for gender in Genders:
    for income in Income_Cat:
        # Create a mask to filter records with missing education level for the current group
        missing_education_for_group_mask = (
            (credit_clean['Income_Category'] == income) &
            (credit_clean['Gender'] == gender) &
            (credit_clean['Education_Level'].isnull())
        )

        # Count how many records are missing Education_Level for this group
        num_missing_education = missing_education_for_group_mask.sum()

        # Proceed only if there are missing values to impute
        if num_missing_education > 0:
            
            # Retrieve known education distributions for the current group
            filter_probabilities = customer[
                (customer['Income_Category'] == income) &
                (customer['Gender'] == gender)
            ]
            
            # Extract possible education levels and their corresponding probabilities  
            education_choices = filter_probabilities['Education_Level'].tolist()
            education_probabilities = filter_probabilities['client_ratio'].tolist()

            # Ensure valid choices exist and probabilities are meaningful
            if education_choices and sum(education_probabilities) > 0: # Ensure probabilities are not all zero
                # Normalise probabilities to sum to 1 if needed (avoiding floating-point inconsistencies)
                prob_sum = sum(education_probabilities)
                if prob_sum != 1.0:
                    education_probabilities = [p / prob_sum for p in education_probabilities]

                # Randomly assign Education_Level values based on observed probabilities
                random_imputations = np.random.choice(
                    a=education_choices,
                    p=education_probabilities,
                    size=num_missing_education
                )

                # Assign imputed values directly using the mask 
                credit_clean.loc[missing_education_for_group_mask, 'Education_Level'] = random_imputations

                print(f"\nSuccessfully imputed {num_missing_education} for income range '{income}' and gender '{gender}' missing 'Education_Level' values.")
            else:
                print(f"\nSkipping imputation for income range '{income}' and gender '{gender}': No valid education choices or probabilities found in 'client' DataFrame.")
        else:
            print(f"\nNo missing 'Education_Level' for income range '{income}' and gender '{gender}' values found to impute.")

# Display updated dataframe structure after imputations 
print("\n--- data_test_imputed after imputation ---")
credit_clean.info()


Successfully imputed 120 for income range '$120K +' and gender 'M' missing 'Education_Level' values.

Successfully imputed 219 for income range '$80K - $120K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 210 for income range '$60K - $80K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 114 for income range '$40K - $60K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 39 for income range 'Less than $40K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 5 for income range 'Unknown' and gender 'M' missing 'Education_Level' values.

No missing 'Education_Level' for income range '$120K +' and gender 'F' values found to impute.

No missing 'Education_Level' for income range '$80K - $120K' and gender 'F' values found to impute.

No missing 'Education_Level' for income range '$60K - $80K' and gender 'F' values found to impute.

Successfully imputed 155 for income range '$40K - $60K' and gender 'F' mi

In [16]:
# Review the updated distribution of Education_Level after imputation to verify replacement of missing values  
credit_clean['Education_Level'].value_counts()

Education_Level
Graduate         3695
High School      2364
Uneducated       1748
College          1191
Post-Graduate     605
Doctorate         524
Name: count, dtype: int64

Comparison Between My Clean Dataset and the One Used in the Project

Below is a comparison of education level counts between my cleaned dataset and the Dataquest dataset:

| Education Level | My Data | Dataquest Data | Difference |
|-----------------|---------|----------------|------------|
| Graduate        | 3,657   | 3,685          | −28        |
| High School     | 2,372   | 2,351          | 21         |
| Uneducated      | 1,774   | 1,755          | 19         |
| College         | 1,189   | 1,192          | −3         |
| Post Graduate   | 602     | 616            | −14        |
| Doctorate       | 533     | 528            | 5          |

Although the individual values vary slightly each time the code is re-run due to the randomness in imputation, the total difference consistently sums to zero. This indicates the imputation maintains the overall data volume while slightly adjusting the category distribution.

### Cleaning the Income Category

In [17]:
# Remove dollar signs and 'K' from Income_Category values to standardise format for analysis
minus_kd=credit_clean['Income_Category'].apply(
    lambda x: x.replace('K', '').replace('$', '')
)
minus_kd_df = pd.DataFrame(minus_kd)

In [18]:
# Map cleaned income values to descriptive income level categories for clearer interpretation
Income_mapping = {'Less than 40' : 'Survival Income',
                  '40 - 60' : 'Basic Needs Income',
                  '60 - 80' : 'Discretionary Income',
                  '80 - 120' : 'Significant Income',
                  '120 +' : 'Abundance Income',
                  'Unknown' : 'Unknown'
                 }
minus_kd_df['Income_level'] = minus_kd_df['Income_Category'].map(Income_mapping)

In [19]:
# Preview the first five rows to verify cleaned and mapped income levels
minus_kd_df.head()

Income_Category          Income_level
0         60 - 80  Discretionary Income
1    Less than 40       Survival Income
2        80 - 120    Significant Income
3    Less than 40       Survival Income
4         60 - 80  Discretionary Income

In [20]:
# Insert the mapped Income_level column into credit_clean at position 7 for consistent column ordering
credit_clean.insert(7, 'Income_level', minus_kd_df['Income_level'])

In [21]:
# Count occurrences of each Income_level category and reset the index for a tidy DataFrame view
credit_clean['Income_level'].value_counts().reset_index()

Income_level  count
0       Survival Income   3561
1    Basic Needs Income   1790
2    Significant Income   1535
3  Discretionary Income   1402
4               Unknown   1112
5      Abundance Income    727

In [22]:
# Remove the Income_Category column as it is no longer needed after mapping to Income_level
credit_clean = credit_clean.drop("Income_Category", axis=1)

In [23]:
# Explore distribution of Education_Level and Gender for customers with unknown Income_level,  
# count combinations, tidy the output, and sort by Education_Level for easy interpretation

(credit_clean[credit_clean['Income_level'] == 'Unknown' ]
 [['Education_Level','Gender']].value_counts().reset_index().sort_values(by=['Education_Level']))

Education_Level Gender  count
3          College      F    119
11         College      M      3
4        Doctorate      F     81
9        Doctorate      M      5
0         Graduate      F    363
6         Graduate      M     23
1      High School      F    251
7      High School      M      9
5    Post-Graduate      F     47
10   Post-Graduate      M      5
2       Uneducated      F    199
8       Uneducated      M      7

In [24]:
# Replace 'Unknown' in Income_level with NaN to treat it as missing data for further analysis or imputation
credit_clean['Income_level'].replace('Unknown', np.NaN, inplace = True)
credit_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           10127 non-null  object 
 5   Marital_Status            10127 non-null  object 
 6   Income_level              9015 non-null   object 
 7   Months_on_book            10127 non-null  int64  
 8   Total_Relationship_Count  10127 non-null  int64  
 9   Months_Inactive_12_mon    10127 non-null  int64  
 10  Credit_Limit              10127 non-null  float64
 11  Total_Trans_Amt           10127 non-null  int64  
 12  Total_Trans_Ct            10127 non-null  int64  
 13  Avg_Utilization_Ratio     10127 non-null  float64
dtypes: flo

In [25]:
# Calculate income level distribution by Education_Level and Gender to guide imputation for unknown income entries

# Define education levels and gender groups for iteration 
Education_Level = ['College', 'Doctorate', 'Graduate', 'High School', 'Post-Graduate', 'Uneducated']
Genders = ['M', 'F']

# List to collect processed income distribution data  
all_customer_data = []

# Iterate over each gender group  
for Gender in Genders: # Outer loop for Gender
    for ed_level in Education_Level: # Inner loop for Income Category
        # Filter for records with the current education level and gender, excluding unknown income levels  
        filtered_data = credit_clean[
            (credit_clean['Education_Level'] == ed_level) &
            (credit_clean['Income_level'] != 'Unknown') &
            (credit_clean['Gender'] == Gender) # Use the Gender from the outer loop
        ]

        # Count total customers matching current filter criteria  
        total_customer = filtered_data.shape[0]
        
        if total_customer > 0:
            # Count frequency of each income level in the filtered subset
            data = filtered_data['Income_level'].value_counts().reset_index()
            data.columns = ['Income_level', 'count']

            # Append contextual attributes
            data['Education_Level'] = ed_level
            data['Gender'] = Gender 

            # Calculate the proportion of customers in each income level
            data['client_ratio'] = data['count'] / total_customer

            all_customer_data.append(data)
        else:
            # Notify if no qualifying customers are found for the current group
            print(f"No {Gender} customers with known education level found for Income Category: {Income}")

# Combine all the income distribution data into a single DataFrame
customer = pd.concat(all_customer_data, ignore_index=True)

# Display the final grouped and weighted income distribution
customer

Income_level  count Education_Level Gender  client_ratio
0     Significant Income    205         College      M      0.360281
1   Discretionary Income    156         College      M      0.274165
2       Abundance Income     90         College      M      0.158172
3     Basic Needs Income     84         College      M      0.147627
4        Survival Income     31         College      M      0.054482
5   Discretionary Income     70       Doctorate      M      0.308370
6     Significant Income     67       Doctorate      M      0.295154
7       Abundance Income     43       Doctorate      M      0.189427
8     Basic Needs Income     30       Doctorate      M      0.132159
9        Survival Income     12       Doctorate      M      0.052863
10    Significant Income    559        Graduate      M      0.327859
11  Discretionary Income    490        Graduate      M      0.287390
12    Basic Needs Income    272        Graduate      M      0.159531
13      Abundance Income    242        Graduate      M      0.141935
14       Survival Income    119        Graduate      M      0.069795
15  Discretionary Income    365     High School      M      0.312500
16    Significant Income    360     High School      M      0.308219
17    Basic Needs Income    193     High School      M      0.165240
18      Abundance Income    180     High School      M      0.154110
19       Survival Income     61     High School      M      0.052226
20  Discretionary Income     92   Post-Graduate      M      0.315068
21    Significant Income     90   Post-Graduate      M      0.308219
22    Basic Needs Income     61   Post-Graduate      M      0.208904
23      Abundance Income     34   Post-Graduate      M      0.116438
24       Survival Income     10   Post-Graduate      M      0.034247
25    Significant Income    254      Uneducated      M      0.314356
26  Discretionary Income    229      Uneducated      M      0.283416
27      Abundance Income    138      Uneducated      M      0.170792
28    Basic Needs Income    136      Uneducated      M      0.168317
29       Survival Income     44      Uneducated      M      0.054455
30       Survival Income    378         College      F      0.607717
31    Basic Needs Income    125         College      F      0.200965
32       Survival Income    169       Doctorate      F      0.569024
33    Basic Needs Income     47       Doctorate      F      0.158249
34       Survival Income   1245        Graduate      F      0.625628
35    Basic Needs Income    382        Graduate      F      0.191960
36       Survival Income    716     High School      F      0.598662
37    Basic Needs Income    229     High School      F      0.191472
38       Survival Income    193   Post-Graduate      F      0.616613
39    Basic Needs Income     73   Post-Graduate      F      0.233227
40       Survival Income    583      Uneducated      F      0.620213
41    Basic Needs Income    158      Uneducated      F      0.168085

In [26]:
# Impute missing Income_level values based on known distributions by Education_Level and Gender

# Define education levels and gender groups for iteration 
Education_Level = ['College', 'Doctorate', 'Graduate', 'High School', 'Post-Graduate', 'Uneducated']
Genders = ['M', 'F']

# Iterate through gender and education levels combinations 
for gender in Genders:
    for ed_level in Education_Level:
        # Identify records with missing income level for the current education level and gender  
        missing_income_for_group_mask = (
            (credit_clean['Education_Level'] == ed_level) &
            (credit_clean['Gender'] == gender) &
            (credit_clean['Income_level'].isnull())
        )

        # Total missing entries to impute 
        num_missing_income = missing_income_for_group_mask.sum()

        # Proceed only if there are missing income records to fill
        if num_missing_income > 0:
            # Retrieve the income distribution for this group from the 'customer' DataFrame  
            filter_probabilities = customer[
                (customer['Education_Level'] == ed_level) &
                (customer['Gender'] == gender)
            ]

            income_choices = filter_probabilities['Income_level'].tolist()
            income_probabilities = filter_probabilities['client_ratio'].tolist()
           
            # Ensure we have valid income categories and non-zero probabilities
            if income_choices and sum(income_probabilities) > 0: # Ensure probabilities are not all zero
                # Normalise probabilities to ensure they sum to 1
                prob_sum = sum(income_probabilities)
                if prob_sum != 1.0:
                    income_probabilities = [p / prob_sum for p in income_probabilities]

                # Generate random imputed values using observed income distribution
                random_imputations = np.random.choice(
                    a=income_choices,
                    p=income_probabilities,
                    size=num_missing_income
                )

                # Assign the imputed values back to the appropriate rows in credit_clean  
                credit_clean.loc[missing_income_for_group_mask, 'Income_level'] = random_imputations

                print(f"\nSuccessfully imputed {num_missing_income} for education level '{ed_level}' and gender '{gender}' missing 'Income_Level' values.")
            else:
                print(f"\nSkipping imputation for education level '{ed_level}' and gender '{gender}': No valid education choices or probabilities found in 'client' DataFrame.")
        else:
            print(f"\nNo missing 'Income_level' for education level '{ed_level}' and gender '{gender}' values found to impute.")

# Display the structure of the DataFrame after imputation
print("\n--- data_test_imputed after imputation ---")
credit_clean.info()



Successfully imputed 3 for education level 'College' and gender 'M' missing 'Income_Level' values.

Successfully imputed 5 for education level 'Doctorate' and gender 'M' missing 'Income_Level' values.

Successfully imputed 23 for education level 'Graduate' and gender 'M' missing 'Income_Level' values.

Successfully imputed 9 for education level 'High School' and gender 'M' missing 'Income_Level' values.

Successfully imputed 5 for education level 'Post-Graduate' and gender 'M' missing 'Income_Level' values.

Successfully imputed 7 for education level 'Uneducated' and gender 'M' missing 'Income_Level' values.

Successfully imputed 119 for education level 'College' and gender 'F' missing 'Income_Level' values.

Successfully imputed 81 for education level 'Doctorate' and gender 'F' missing 'Income_Level' values.

Successfully imputed 363 for education level 'Graduate' and gender 'F' missing 'Income_Level' values.

Successfully imputed 251 for education level 'High School' and gender 'F' 

In [27]:
# Count occurrences of each Income_level category and reset the index for a well-structured DataFrame  
credit_clean['Income_level'].value_counts().reset_index()

Income_level  count
0       Survival Income   4373
1    Basic Needs Income   2052
2    Significant Income   1550
3  Discretionary Income   1420
4      Abundance Income    732

### Changing Columns name to matches with dataquest data

In [28]:
# Rename dataset columns to align with Dataquest naming conventions,  
# except for 'income_level', which replaces estimated income with a more appropriate classification  

credit_clean.rename(columns={'CLIENTNUM':'customer_id',
                             'Customer_Age':'age',
                            'Gender' : 'gender',
                            'Dependent_count':'dependent_count',
                            'Education_Level':'education_level',
                            'Marital_Status':'marital_status',
                            'Income_level':'income_level',
                            'Months_on_book':'months_on_book',
                            'Total_Relationship_Count':'total_relationship_count',
                            'Months_Inactive_12_mon':'months_inactive_12_mon',
                            'Credit_Limit':'credit_limit',
                            'Total_Trans_Amt':'total_trans_amount',
                            'Total_Trans_Ct':'total_trans_count',
                            'Avg_Utilization_Ratio':'avg_utilization_ratio'}, inplace=True)

credit_clean.head()

customer_id  age gender  dependent_count education_level marital_status  \
0    768805383   45      M                3     High School        Married   
1    818770008   49      F                5        Graduate         Single   
2    713982108   51      M                3        Graduate        Married   
3    769911858   40      F                4     High School        Unknown   
4    709106358   40      M                3      Uneducated        Married   

           income_level  months_on_book  total_relationship_count  \
0  Discretionary Income              39                         5   
1       Survival Income              44                         6   
2    Significant Income              36                         4   
3       Survival Income              34                         3   
4  Discretionary Income              21                         5   

   months_inactive_12_mon  credit_limit  total_trans_amount  \
0                       1       12691.0                1144   
1                       1        8256.0                1291   
2                       1        3418.0                1887   
3                       4        3313.0                1171   
4                       1        4716.0                 816   

   total_trans_count  avg_utilization_ratio  
0                 42                  0.061  
1                 33                  0.105  
2                 20                  0.000  
3                 20                  0.760  
4                 28                  0.000

In [29]:
# Export the cleaned dataset to a CSV file named 'customer_segmentation.csv',  
# ensuring consistency with Dataquest file naming conventions  
credit_clean.to_csv('customer_segmentation.csv', index = False)